File: backend/src/test/java/com/yigongil/backend/domain/study/PeriodUnitTest.java
package com.yigongil.backend.domain.study;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

class PeriodUnitTest {

    @CsvSource({"12w,WEEK", "57d,DAY"})
    @ParameterizedTest
    void 입력을_받아서_적절히_변환한다(String input, PeriodUnit expectedPeriodUnit) {
        // given
        int expectedNumber = Integer.parseInt(input.substring(0, input.length() - 1));

        // when
        Integer periodNumber = PeriodUnit.getPeriodNumber(input);
        PeriodUnit periodUnit = PeriodUnit.getPeriodUnit(input);

        // then
        assertThat(periodNumber).isEqualTo(expectedNumber);
        assertThat(periodUnit).isEqualTo(expectedPeriodUnit);
    }

    @CsvSource({"12w,12,WEEK", "57d,57,DAY"})
    @ParameterizedTest
    void 있는_데이터로_반환_데이터를_만든다(String expected, int periodNumber, PeriodUnit periodUnit) {
        // given

        // when
        String actual = periodUnit.toStringFormat(periodNumber);

        // then
        assertThat(actual).isEqualTo(expected);
    }
}

File: android/domain/src/test/java/com/created/domain/model/PeriodUnitTest.kt
package com.created.domain.model

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

class PeriodUnitTest {
    @Test
    fun `type이 0이면 PeriodUnit은 DAY이다`() {
        val periodUnit = PeriodUnit.valueOf(0)
        val expected = PeriodUnit.DAY
        assertEquals(expected, periodUnit)
    }

    @Test
    fun `type이 1이면 PeriodUnit은 WEEK이다`() {
        val periodUnit = PeriodUnit.valueOf(1)
        val expected = PeriodUnit.WEEK
        assertEquals(expected, periodUnit)
    }

    @Test
    fun `type이 존재하지 않는다면 예외가 발생한다`() {
        val exception = assertThrows<IllegalArgumentException> {
            throw IllegalArgumentException(ERROR_ARGUMENT)
        }
        assertEquals(ERROR_ARGUMENT, exception.message)
    }

    @Test
    fun `type이 d이면 PeriodUnit은 DAY이다`() {
        val periodUnit = PeriodUnit.valueOf('d')
        val expected = PeriodUnit.DAY
        assertEquals(expected, periodUnit)
    }

    @Test
    fun `type이 w이면 PeriodUnit은 WEEK이다`() {
        val periodUnit = PeriodUnit.valueOf('w')
        val expected = PeriodUnit.WEEK
        assertEquals(expected, periodUnit)
    }

    companion object {
        private const val ERROR_ARGUMENT = "존재하지 않는 타입입니다"
    }
}
